arangod/Graph/GraphManager.cpp

```diff
@@ -227,62 +188,88 @@ Result GraphManager::checkForEdgeDefinitionConflicts(std::map<std::string, EdgeD
     }
+  } else if (!res.is(TRI_ERROR_ARANGO_DATA_SOURCE_NOT_FOUND)) {
+    return res;
+  } else {
+    edgeCollectionsToCreate.emplace(edgeCollName);
   }
 
-  return Result{TRI_ERROR_NO_ERROR};
-}
-
-Result GraphManager::findOrCreateCollectionsByEdgeDefinition(Graph const& graph,
-                                                             EdgeDefinition const& edgeDefinition,
-                                                             bool waitForSync,
-                                                             VPackSlice const options) {
-  std::string const& edgeCollection = edgeDefinition.getName();
-  std::shared_ptr<LogicalCollection> def =
-      getCollectionByName(ctx()->vocbase(), edgeCollection);
-
-  if (def == nullptr) {
-    Result res = createEdgeCollection(edgeCollection, waitForSync, options);
-    if (res.fail()) {
+  for (auto const& vertexColl : edgeDefinition.getFrom()) {
+    std::shared_ptr<LogicalCollection> col;
+    Result res = methods::Collections::lookup(vocbase, vertexColl, col);
+    if (res.ok()) {
+      TRI_ASSERT(col);
+      if (col->isSatellite()) {
+        satellites.emplace(col->name());
+      }
+      existentDocumentCollections.emplace(col);
+    } else if (!res.is(TRI_ERROR_ARANGO_DATA_SOURCE_NOT_FOUND)) {
       return res;
+    } else {
+      if (edgeCollectionsToCreate.find(vertexColl) == edgeCollectionsToCreate.end()) {
+        auto res = ensureVertexShardingMatches(graph, *edgeColl, satellites,
+                                               vertexColl, true);
+        if (res.fail()) {
+          return res;
+        }
+        documentCollectionsToCreate.emplace(vertexColl);
+      }
     }
   }
 
-  std::unordered_set<std::string> vertexCollections;
-
-  // duplicates in from and to shouldn't occur, but are safely ignored here
-  for (auto const& colName : edgeDefinition.getFrom()) {
-    vertexCollections.emplace(colName);
-  }
-  for (auto const& colName : edgeDefinition.getTo()) {
-    vertexCollections.emplace(colName);
-  }
-  for (auto const& colName : vertexCollections) {
-    def = getCollectionByName(ctx()->vocbase(), colName);
-    if (def == nullptr) {
-      Result res = createVertexCollection(colName, waitForSync, options);
-      if (res.fail()) {
-        return res;
+  for (auto const& vertexColl : edgeDefinition.getTo()) {
+    std::shared_ptr<LogicalCollection> col;
+    Result res = methods::Collections::lookup(vocbase, vertexColl, col);
+    if (res.ok()) {
+      TRI_ASSERT(col);
+      if (col->isSatellite()) {
+        satellites.emplace(col->name());
       }
+      existentDocumentCollections.emplace(col);
+    } else if (!res.is(TRI_ERROR_ARANGO_DATA_SOURCE_NOT_FOUND)) {
+      return res;
     } else {
-      auto res = graph.validateCollection(*def.get());
-      if (res.fail()) {
-        return res;
+      if (edgeCollectionsToCreate.find(vertexColl) == edgeCollectionsToCreate.end()) {
+        // We can make this EE Code
```

Comment:
Let's create a Kanban card for that before we merge, so this will not be forgotten.

Reply:
```suggestion
```

Reply:
Good catch, I was talking about the Method called below here, which is already EE.